File: portal/src/components/visual/bracketLayout.ts
// Pure bracket layout math — the React port of the SDK's Compose BracketLayout.kt.
// Same model: round → x (columns), round-1 slot → y (rows), each later match vertically CENTERED
// between the two matches that feed it (its children via nextMatchId), with elbow connectors.
import type { Match } from "../../lib/types";

export interface BracketPos {
  matchId: string;
  x: number;
  y: number;
}

export interface BracketConnector {
  fromId: string;
  toId: string;
}

export interface BracketLayout {
  positions: Record<string, BracketPos>;
  connectors: BracketConnector[];
  width: number;
  height: number;
}

// Card + gap sizing (px). Exported so the BracketView draws cards/connectors with the same numbers.
export interface BracketMetrics {
  cardWidth: number;
  cardHeight: number;
  hGap: number; // horizontal gap between rounds
  vGap: number; // vertical gap between round-1 cards
}

export const DEFAULT_METRICS: BracketMetrics = {
  cardWidth: 200,
  cardHeight: 92,
  hGap: 56,
  vGap: 26,
};

// Computes the left→right bracket layout from the engine's match list. Pure; no DOM.
export function computeBracketLayout(matches: Match[], metrics: BracketMetrics = DEFAULT_METRICS): BracketLayout {
  if (matches.length === 0) return { positions: {}, connectors: [], width: 0, height: 0 };

  const byId: Record<string, Match> = {};
  for (const m of matches) byId[m.id] = m;

  const rounds = Array.from(new Set(matches.map((m) => m.round))).sort((a, b) => a - b);
  const firstRound = rounds[0];

  // Column x for a round (rounds are contiguous in a single-elimination bracket).
  const xForRound = (round: number) => (round - firstRound) * (metrics.cardWidth + metrics.hGap);

  const positions: Record<string, BracketPos> = {};

  // Round 1 (leaves): stack by slot order — the vertical anchor for the whole tree.
  matches
    .filter((m) => m.round === firstRound)
    .sort((a, b) => a.slot - b.slot)
    .forEach((m, index) => {
      positions[m.id] = { matchId: m.id, x: xForRound(m.round), y: index * (metrics.cardHeight + metrics.vGap) };
    });

  // Later rounds: a parent's y is the mean of its already-placed children's y (parent-centering).
  for (const round of rounds.slice(1)) {
    matches
      .filter((m) => m.round === round)
      .sort((a, b) => a.slot - b.slot)
      .forEach((m) => {
        const childYs = matches
          .filter((c) => c.nextMatchId === m.id)
          .map((c) => positions[c.id]?.y)
          .filter((y): y is number => y !== undefined);
        const y =
          childYs.length > 0
            ? childYs.reduce((a, b) => a + b, 0) / childYs.length
            : m.slot * (metrics.cardHeight + metrics.vGap);
        positions[m.id] = { matchId: m.id, x: xForRound(round), y };
      });
  }

  // One connector per match that advances (winner feeds nextMatchId).
  const connectors: BracketConnector[] = matches
    .filter((m) => m.nextMatchId != null && byId[m.nextMatchId])
    .map((m) => ({ fromId: m.id, toId: m.nextMatchId as string }));

  const width = rounds.length * metrics.cardWidth + (rounds.length - 1) * metrics.hGap;
  const maxY = Math.max(0, ...Object.values(positions).map((p) => p.y));
  const height = maxY + metrics.cardHeight;
  return { positions, connectors, width, height };
}

File: kotlin/sdk/src/main/kotlin/com/tournamentkit/sdk/ui/BracketLayout.kt
package com.tournamentkit.sdk.ui

import com.tournamentkit.shared.Match

// A computed position for one match card, in abstract pixels (Dp values are applied by the composable).
data class BracketPos(val matchId: String, val x: Float, val y: Float)

// A connector from a child match's right edge to its parent (nextMatchId) match's left edge.
data class BracketConnector(val fromId: String, val toId: String)

// The full computed layout: where each card goes, the connectors, and the overall canvas size.
data class BracketLayout(
    val positions: Map<String, BracketPos>,
    val connectors: List<BracketConnector>,
    val width: Float,
    val height: Float
)

// Sizing inputs for the layout math (kept as plain numbers so this is pure and unit-testable).
data class BracketMetrics(
    val cardWidth: Float,
    val cardHeight: Float,
    val hGap: Float,   // horizontal gap between rounds
    val vGap: Float    // vertical gap between round-1 cards
)

// Pure layout math mapping the match list to x-by-round, y-by-slot, with each later match centered between its two feeders.
fun computeBracketLayout(matches: List<Match>, metrics: BracketMetrics): BracketLayout {
    if (matches.isEmpty()) return BracketLayout(emptyMap(), emptyList(), 0f, 0f)

    val byId = matches.associateBy { it.id }
    val rounds = matches.map { it.round }.distinct().sorted()

    // Column x for a round (rounds are 1-based and contiguous in a single-elimination bracket).
    fun xForRound(round: Int): Float = (round - rounds.first()) * (metrics.cardWidth + metrics.hGap)

    val positions = HashMap<String, BracketPos>()

    // Round 1 (the leaves): stack by slot order so the whole tree has a stable vertical anchor.
    val firstRound = matches.filter { it.round == rounds.first() }.sortedBy { it.slot }
    firstRound.forEachIndexed { index, m ->
        positions[m.id] = BracketPos(m.id, xForRound(m.round), index * (metrics.cardHeight + metrics.vGap))
    }

    // Later rounds: a parent's y is the mean of its already-placed children's y (parent-centering).
    for (round in rounds.drop(1)) {
        val inRound = matches.filter { it.round == round }.sortedBy { it.slot }
        for (m in inRound) {
            val childYs = matches
                .filter { it.nextMatchId == m.id }
                .mapNotNull { positions[it.id]?.y }
            // Fall back to slot anchoring if (unexpectedly) no child is placed yet.
            val y = if (childYs.isNotEmpty()) childYs.average().toFloat()
            else m.slot * (metrics.cardHeight + metrics.vGap)
            positions[m.id] = BracketPos(m.id, xForRound(round), y)
        }
    }

    // One connector per match that advances somewhere (winner feeds nextMatchId).
    val connectors = matches
        .filter { it.nextMatchId != null && byId.containsKey(it.nextMatchId) }
        .map { BracketConnector(it.id, it.nextMatchId!!) }

    val width = (rounds.size * metrics.cardWidth) + ((rounds.size - 1) * metrics.hGap)
    val height = (positions.values.maxOfOrNull { it.y } ?: 0f) + metrics.cardHeight
    return BracketLayout(positions, connectors, width, height)
}
